### DISCUSSION TITLE
How to configure ghostscript on macos

### DISCUSSION THREAD

**Discussion by nielsbosma** [dlemstra/Magick.NET#1207] [Help]
I'm using this ghostscript build from anaconda [osx-64/ghostscript-9.54.0-h96cf925_2.tar.bz2](https://anaconda.org/conda-forge/ghostscript/9.54.0/download/osx-64/ghostscript-9.54.0-h96cf925_2.tar.bz2). 

I don't want to install ghostscript on the machine so I unzip this to ~/.ghostscript/

In my C# code I then set: 

MagickNET.SetGhostscriptDirectory("~/.ghostscript/bin");

I'm then getting this error:

```ImageMagick.MagickDelegateErrorException: FailedToExecuteCommand `'gs' -sstdout=%stderr -dQUIET -dSAFER -dBATCH -dNOPAUSE -dNOPROMPT -dMaxBitmap=500000000 -dAlignToPixels=0 -dGridFitTT=2 '-sDEVICE=pngalpha' -dTextAlphaBits=4 -dGraphicsAlphaBits=4 '-r72x72' -dFirstPage=1 -dLastPage=1 -dEPSCrop  '-sOutputFile=/var/folders/z2/l2wtfwl949b0zj6t8t_vj2p40000gn/T/magick-rzApJ8zxK2Rb2```

On Windows this works fine.

The build seems to be fine when running directly in the CLI.

**Comment by dlemstra** [comment]
I don't understand what you mean by this statement: 

> The build seems to be fine when running directly in the CLI.

**Reply by nielsbosma** [reply]
What I mean is that the ghostscript library in ~/.ghostscript/bin works fine from the CLI.

**Comment by dlemstra** [comment]
The `MagickNET.SetGhostscriptDirectory` method only works on Windows. I will update the documentation to make this more clear. I am also wondering if I should throw an exception on a non windows platform for those that don't read the documentation. I don't know how you can do this but you should add `~/.ghostscript/bin` to the path of the user that is running the application.

**Reply by nielsbosma** [reply]
In my scenario I'm not allowed to install anything on the users computer (including changing the PATH). Is there any way to set the PATH temporarily the the IM process?

**Reply by dlemstra** [reply]
In the native library that is being used there is a method to set an environment variable to a specific value. But that would require you to call the native code yourself and this does not come with any guarantees that this will continue to work in the future. And currently there is no method to get an environment variable because you might need to get the current path value.

I am now thinking of an other solution where you could use an method to set the path to the `gs` executable (e.g. `~/.ghostscript/bin/gs`) but that will probably also require me to make some changes inside ImageMagick. I will take a look at that this weekend.

**Reply by nielsbosma** [reply]
Thanks! I have the same issue with fontconfig. 🙏

**Reply by dlemstra** [reply]
What kind of issue are you having with fontconfig?

**Reply by nielsbosma** [reply]
I get:

//Fontconfig error: Cannot load default config file
//Get the exception: ImageMagick.MagickTypeErrorException: UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1538

**Reply by nielsbosma** [reply]
[ImageMagickTest.zip](https://github.com/dlemstra/Magick.NET/files/9034864/ImageMagickTest.zip)

I made a simple possible console app to show the issues with fontconfig and fs.

**Reply by dlemstra** [reply]
Thanks for the console application but I don't own a mac so I cannot reproduce this locally. Have you tried to use `MagickNET.SetFontConfigDirectory`?

**Reply by nielsbosma** [reply]
Yes, I've tried that. I'd be happy to sponsor with a mac cloud VM if that helps?

**Reply by nielsbosma** [reply]
I decided I don't need Ghostscript in the end. But would be very good if there was any way to get fontconfig working.

**Reply by dlemstra** [reply]
Why is the `MagickNET.SetFontConfigDirectory` method not working? Did you put the required files in that folder?

**Reply by nielsbosma** [reply]
I keep getting Fontconfig error: Cannot load default config file
ImageMagick.MagickTypeErrorException: UnableToReadFont `helvetica' @ error/annotate.c/RenderFreetype/1538
   at ImageMagick.NativeInstance.CheckException(IntPtr exception, IntPtr result) in /_/src/Magick.NET/Native/NativeInstance.cs:line 57
   at ImageMagick.MagickImage.NativeMagickImage.ReadFile(IMagickSettings`1 settings) in /_/src/Magick.NET/Native/MagickImage.cs:line 7159
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings, Boolean ping) in /_/src/Magick.NET/MagickImage.cs:line 7344
   at ImageMagick.MagickImage.Read(String fileName, IMagickReadSettings`1 readSettings) in /_/src/Magick.NET/MagickImage.cs:line 5124
   at ImageMagick.MagickImage.Read(String fileName) in /_/src/Magick.NET/MagickImage.cs:line 5085
   at ImageMagick.MagickImage..ctor(String fileName) in /_/src/Magick.NET/MagickImage.cs:line 198
   at ImageMagickTest.Program.TestWithText() in /Users/filestar/Desktop/ImageMagickTest/Program.cs:line 55
   at ImageMagickTest.Program.Main(String[] args) in /Users/filestar/Desktop/ImageMagickTest/Program.cs:line 23

What is the required files?

I download this https://anaconda.org/conda-forge/fontconfig/2.14.0/download/osx-64/fontconfig-2.14.0-h676cef8_0.tar.bz2

Unzip it and set:

 MagickNET.SetFontConfigDirectory(Path.Combine(fcPath, "bin"));

(I've also tried without the "bin" part)

**Comment by nielsbosma** [comment]
FYI: I got it working by setting:
MagickNET.SetFontConfigDirectory(Path.Combine(fcPath, "etc", "fonts"));

My issue now is that fontconfig spits out a bunch of warnings on the console: 

Fontconfig warning: "/Users/filestar/.config/com.filestar.macos/resources/debug/fontconfig/2.14.0/etc/fonts/fonts.conf", line 5: unknown element "description"
Fontconfig warning: "/Users/filestar/.config/com.filestar.macos/resources/debug/fontconfig/2.14.0/etc/fonts/conf.d/10-hinting-slight.conf", line 4: unknown element "description"
Fontconfig warning: "/Users/filestar/.config/com.filestar.macos/resources/debug/fontconfig/2.14.0/etc/fonts/conf.d/10-scale-bitmap-fonts.conf", line 4: unknown element "description"
Fontconfig warning: "/Users/filestar/.config/com.filestar.macos/resources/debug/fontconfig/2.14.0/etc/fonts/conf.d/11-lcdfilter-default.conf", line 4: unknown element "description"

Can these be turned off?

/Niels

**Reply by dlemstra** [reply]
You will need to patch your `fonts.conf` file and remove the unknown element `description`.
